drivers/flash/flash_mcux_flexspi_hyperflash.c

```diff
@@ -0,0 +1,609 @@
+		void *buffer, size_t len)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	uint8_t *src = memc_flexspi_get_ahb_address(data->controller,
+			config->port,
+			offset);
+	if (!src) {
+		return -EINVAL;
+	}
+
+	memcpy(buffer, src, len);
+
+	return 0;
+}
+
+static int flash_flexspi_hyperflash_write(const struct device *dev, off_t offset,
+		const void *buffer, size_t len)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	size_t size = len;
+	uint8_t *src = (uint8_t *)buffer;
+	unsigned int key = 0;
+	int i;
+	int ret;
+
+	uint8_t *dst = memc_flexspi_get_ahb_address(data->controller,
+			config->port,
+			offset);
+	if (!dst) {
+		return -EINVAL;
+	}
+
+	if (memc_flexspi_is_running_xip(data->controller)) {
+		key = irq_lock();
+	}
+
+	if ((offset % SPI_HYPERFLASH_PAGE_SIZE) != 0) {
+		LOG_ERR("not page aligned");
+	}
+
+	while (len) {
+		/* If the offset isn't a multiple of the NOR page size, we first need
+		 * to write the remaining part that fits, otherwise the write could
+		 * be wrapped around within the same page
+		 */
+		i = MIN(SPI_HYPERFLASH_PAGE_SIZE - (offset %
+					SPI_HYPERFLASH_PAGE_SIZE), len);
+
+		ret = flash_flexspi_hyperflash_write_enable(dev, offset);
+		if (ret != 0) {
+			LOG_ERR("failed to enable write");
+			break;
+		}
+
+		ret = flash_flexspi_hyperflash_page_program(dev, offset, src, i);
+		if (ret != 0) {
+			LOG_ERR("failed to write");
+			break;
+		}
+
+		ret = flash_flexspi_hyperflash_wait_bus_busy(dev);
+		if (ret != 0) {
+			LOG_ERR("failed to wait bus busy");
+			break;
+		}
+
+		/* Do software reset. */
+		memc_flexspi_reset(data->controller);
+		src += i;
+		offset += i;
+		len -= i;
+	}
+
+	if (memc_flexspi_is_running_xip(data->controller)) {
+		irq_unlock(key);
+	}
+
+#ifdef CONFIG_HAS_MCUX_CACHE
+	DCACHE_InvalidateByRange((uint32_t) dst, size);
+#endif
+
+	return 0;
+}
+
+static int flash_flexspi_hyperflash_erase(const struct device *dev, off_t offset, size_t size)
+{
+	const struct flash_flexspi_hyperflash_config *config = dev->config;
+	struct flash_flexspi_hyperflash_data *data = dev->data;
+	flexspi_transfer_t transfer;
+	int ret;
+	int i;
+	unsigned int key = 0;
+	int num_sectors = size / SPI_HYPERFLASH_SECTOR_SIZE;
+	uint8_t *dst = memc_flexspi_get_ahb_address(data->controller,
+			config->port,
+			offset);
+
+	if (!dst) {
+		return -EINVAL;
+	}
+
+	if (offset % SPI_HYPERFLASH_SECTOR_SIZE) {
+		LOG_ERR("Invalid offset");
+		return -EINVAL;
+	}
+
+	if (size % SPI_HYPERFLASH_SECTOR_SIZE) {
+		LOG_ERR("Invalid offset");
+		return -EINVAL;
+	}
+
+	if (memc_flexspi_is_running_xip(data->controller)) {
+		key = irq_lock();
+	}
+
+	for (i = 0; i < num_sectors; i++) {
+		ret = flash_flexspi_hyperflash_write_enable(dev, offset);
+		if (ret != 0) {
+			LOG_ERR("failed to write_enable");
+			return ret;
```

Comment:
Same here as with the previous function, do not return with the `irq_lock` active.